Scraper for CI/CD

aidnurs opened the pull request:
Closes #217 

--- 

# Squash migrations

After squashing migrations as was done in #232 the following output was generated using `diff` of `backend/schema_dump/before_squash_schema.sql` and `backend/schema_dump/after_squash_schema.sql` files. As you can see the only difference is the naming of the plants table and the order of primary and unique constraints, which is expected. Therefore, we can proceed with squashed migrations further.

```
424,425c424,425
<     CONSTRAINT plant_detail_heat_zone_check CHECK (((heat_zone IS NULL) OR ((heat_zone >= 0) AND (heat_zone <= 13)))),
<     CONSTRAINT plant_detail_preferable_permaculture_zone_check CHECK (((preferable_permaculture_zone IS NULL) OR ((preferable_permaculture_zone >= '-1'::integer) AND (preferable_permaculture_zone <= 6))))
---
>     CONSTRAINT plants_heat_zone_check CHECK (((heat_zone IS NULL) OR ((heat_zone >= 0) AND (heat_zone <= 13)))),
>     CONSTRAINT plants_preferable_permaculture_zone_check CHECK (((preferable_permaculture_zone IS NULL) OR ((preferable_permaculture_zone >= '-1'::integer) AND (preferable_permaculture_zone <= 6))))
432c432
< -- Name: plant_detail_id_seq; Type: SEQUENCE; Schema: public; Owner: permaplant
---
> -- Name: plants_id_seq; Type: SEQUENCE; Schema: public; Owner: permaplant
435c435
< CREATE SEQUENCE public.plant_detail_id_seq
---
> CREATE SEQUENCE public.plants_id_seq
444c444
< ALTER TABLE public.plant_detail_id_seq OWNER TO permaplant;
---
> ALTER TABLE public.plants_id_seq OWNER TO permaplant;
447c447
< -- Name: plant_detail_id_seq; Type: SEQUENCE OWNED BY; Schema: public; Owner: permaplant
---
> -- Name: plants_id_seq; Type: SEQUENCE OWNED BY; Schema: public; Owner: permaplant
450c450
< ALTER SEQUENCE public.plant_detail_id_seq OWNED BY public.plants.id;
---
> ALTER SEQUENCE public.plants_id_seq OWNED BY public.plants.id;
503c503
< ALTER TABLE ONLY public.plants ALTER COLUMN id SET DEFAULT nextval('public.plant_detail_id_seq'::regclass);
---
> ALTER TABLE ONLY public.plants ALTER COLUMN id SET DEFAULT nextval('public.plants_id_seq'::regclass);
522c522
< -- Name: plants plant_detail_pkey; Type: CONSTRAINT; Schema: public; Owner: permaplant
---
> -- Name: plants plant_detail_unique_name_key; Type: CONSTRAINT; Schema: public; Owner: permaplant
526c526
<     ADD CONSTRAINT plant_detail_pkey PRIMARY KEY (id);
---
>     ADD CONSTRAINT plant_detail_unique_name_key UNIQUE (unique_name);
530c530
< -- Name: plants plant_unique_name_key; Type: CONSTRAINT; Schema: public; Owner: permaplant
---
> -- Name: plants plants_pkey; Type: CONSTRAINT; Schema: public; Owner: permaplant
534c534
<     ADD CONSTRAINT plant_unique_name_key UNIQUE (unique_name);
---
>     ADD CONSTRAINT plants_pkey PRIMARY KEY (id);
```


## TODO list:
- [x] external_url
- [x] external_id
- [x] Please execute Wikipedia queries on all plants to get German names.
- [x] medicinal and medicinal_use should be merged. Exchange "true" from permapeople if there is an entry in practicalplants. Otherwise preference is on permapeople.
- [x] fix backend
- [x] document structs in rust files

Individual problems:
- [x] Daucus carota L. ssp. sativus --> Daucus carota sativus
- [x] Petroselinum crispum ssp. tuberosum --> Petroselinum crispum tuberosum
- [x] Papaver somnif. var. paeonifl. --> Papaver somniferum paeoniflorum
- [x] Alcea rosea fl. pl. --> Alcea rosea flore pleno
- [x] Campanula lat. macr. --> Campanula latifolia macrantha
- [x] Malva sylvestris ssp. maur. --> Malva sylvestris mauritiana

<!--
Check relevant points but **please do not remove entries**.
-->

## Basics

<!--
These points need to be fulfilled for every PR.
-->

- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildserver is happy.
- [ ] The PR is rebased with current master.

<!--
If you have any troubles fulfilling these criteria, please write about the trouble as comment in the PR.
We will help you, but we cannot accept PRs that do not fulfill the basics.
-->

## Checklist

<!--
For documentation fixes, spell checking, and similar none of these points below need to be checked.
-->

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed all affected decisions
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I mentioned [every code or binary](/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)

## Review

<!--
Reviewers should check the following.
-->

- [ ] Documentation is conforming to [our Documentation Guidelines](/doc/documentation.md)
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to our Coding Guidelines
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

Benjaminimal commented:
It's a little hard to review this PR if the actual changes and reformatting are not separated by commits.
Some things I noted.
1. If https://github.com/ElektraInitiative/PermaplanT/pull/232 makes it into master you would need to update this statement because the constraint name changes. I can also adjust it over there if you prefer that.
https://github.com/ElektraInitiative/PermaplanT/blob/scraper-ci/scraper/src/insert_plant_details.js#L94-L101
2. I can remember Markus saying in the meeting today that he would prefer all inserts would happen within a single transaction.

markus2330 reviewed (changes requested):
I am not sure what the purpose of this PR is. Should it fix everything in #217?

aidnurs commented:
Hey @markus2330, 

this PR is not finalized yet, nevertheless, I would appreciate a review since there are many files affected by this PR. I'm currently looking through the data and searching for erroneous and inconsistent values, so feel free to list problems here as well!

Thank you!

markus2330 commented on scraper/doc/permapeople_columns.md:
```diff
@@ -0,0 +1,172 @@
+
+| PracticalPlants column name | Permapeople name    | New column name       |
+| --------------------------- | ------------------- | --------------------- |
+| binomial_name               | scientific_name     | scientific_name       |
```
Isn't this unique_name? I think it is better if we document this in the Rust struct. Or is this for Yvonne and you copy it to Rust later?

aidnurs replied:
I created these mapping docs for myself during the merging and before putting them into the struct docs. But I'm currently stuck with Rust and will ask for help in the chat. So I left them for now and can delete afterwards

markus2330 replied:
Ok, then it is okay. I probably cannot give you useful feedback if you don't tell me what is finished and what I should review.

Benjaminimal commented:
> Locally, cargo run hangs forever on the diesel step without any error indications.

For how long did you leave it running?

From https://docs.rs/diesel/latest/diesel/macro.table.html
> Note that enabling 64 column tables or larger will substantially increase the compile time of Diesel.

Benjaminimal commented:
I changed the naming of the `1000_seed_weight` columns to `seed_weight`.
Seems like this caused some issues.

```
cargo run
   Compiling backend v0.1.0 (/home/benjaminimal/Projects/uni/PermaplanT/backend)
error: recursion limit reached while expanding `__static_cond!`
   --> src/schema.rs:73:1
    |
73  | / diesel::table! {
74  | |     use diesel::sql_types::*;
75  | |     use super::sql_types::SoilPh;
76  | |     use super::sql_types::SoilTexture;
...   |
204 | |     }
205 | | }
    | |_^
    |
    = help: consider increasing the recursion limit by adding a `#![recursion_limit = "512"]` attribute to your crate (`backend`)
    = note: this error originates in the macro `$crate::static_cond` which comes from the expansion of the macro `diesel::table` (in Nightly builds, run with -Z macro-backtrace for more info)

error: could not compile `backend` due to previous error
```

After increasing the recursion depth in `main.rs` some more errors popped up.
You haven't hit the column limit but it seems like the generated table has an invalid type for all the traits to work with it.
It might not be avoidable but I would prefer it if we don't have to use the 128-column feature for the sake of build times and complexity.

https://pastebin.com/5yFvaqSM

kitzbergerg commented on backend/src/model/entity.rs:
```diff
@@ -75,6 +78,70 @@ pub struct Plants {
+    pub required_quantity_of_seeds_de: Option<String>,
+    pub required_quantity_of_seeds_en: Option<String>,
+    pub seed_planting_depth_de: Option<String>,
+    pub _1000_seed_weight_de: Option<String>,
+    pub _1000_seed_weight_en: Option<String>,
+    pub _1000_seed_weight: Option<String>,
```
You have to rename the tables to something like `seed_weight_1000_de` (in the database/migration as well). 
In Rust it is not allowed to start variable names with numbers and if you look in the generated `schema.rs` file there are names like `1000_seed_weight_de` which is why it doesn't compile.

kitzbergerg commented:
@aidnurs I'll have a look and push a working version.

kitzbergerg commented:
@aidnurs There still is a mismatch between the database and the types defined in Rust. I couldn't find it even after looking for a while. Maybe you have some idea where that error might have occurred. 
It might have to do with wrong types (see my commit, I had to change the range there) or maybe a field has the wrong name.

For the future please try to always get the pipeline green as soon as possible. There are now quite a few changes and it is really hard to find where the error actually originated.

aidnurs commented:
> @aidnurs There still is a mismatch between the database and the types defined in Rust. I couldn't find it even after looking for a while. Maybe you have some idea where that error might have occurred. It might have to do with wrong types (see my commit, I had to change the range there) or maybe a field has the wrong name.
> 
> For the future please try to always get the pipeline green as soon as possible. There are now quite a few changes and it is really hard to find where the error actually originated.

Thank you @kitzbergerg! You gave me the direction, so if it is only the column types left now then it should be easy. I'll let you know once again after I change the types accordingly!

markus2330 commented:
> For the future please try to always get the pipeline green as soon as possible.

Yes, very important tip.

> There are now quite a few changes and it is really hard to find where the error actually originated.

There still is ` git bisect` but I am afraid it won't help here.

> You gave me the direction, so if it is only the column types left now then it should be easy.

Yes and no, there are many possible type errors and many columns... If you can't find it, remove half of the struct and (the same) half of the columns and see if the error is still here. Then you can maybe narrow it down and also get faster compilation times.

aidnurs commented on doc/database/hierarchy.md:
```diff
@@ -191,10 +191,6 @@ Rename:
 
-Bug:
-
-- `provides_forage_for` or `provides_shelter_for`: seems to be a bug in scraper (currently empty but was in practical plants?)
```
These columns were empty in PracticalPlants as well, therefore, removed from the database too

aidnurs commented:
Hey @markus2330,

all the required features are implemented, the plants table is sanitized, and all the erroneous entries are cleaned. The PR is tested and ready to be merged!

markus2330 commented on backend/migrations/2023-04-12-162600_rename_plants_columns/down.sql:
```diff
@@ -0,0 +1,21 @@
+-- This file should undo anything in `up.sql`
```
Please squash all plants SQL files as done in https://github.com/ElektraInitiative/PermaplanT/pull/232 (with check if they are identical to these files here).

Otherwise it is difficult for me to check these SQL files.

aidnurs replied:
Ok, I prepared them and moved to another folder

markus2330 commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,230 @@ use diesel::{Identifiable, Insertable, Queryable};
+    pub unique_name: String,
+    /// The list of the common names of the plant.
+    pub common_name_en: Option<Vec<Option<String>>>,
+    /// The list of the common names of the plant in german. Fetched from Wikidata API if not present in the source datasets.
```
I would prefer if they are always fetched (and added to the array if not already there).

```suggestion
    /// The list of the common names of the plant in German. Fetched from Wikidata API if not present in the source datasets.
```

markus2330 commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,230 @@ use diesel::{Identifiable, Insertable, Queryable};
+    ///
     pub ecosystem_niche: Option<String>,
-    pub root_zone_tendancy: Option<RootZoneTendancy>,
+    /// Enum value.
```
Please remove such comments (it can be seen from the type anyway). Important is which columns you fetched from where, some non-obvious mapping you might have done etc.

markus2330 commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,230 @@ use diesel::{Identifiable, Insertable, Queryable};
+    /// Date value fetched from PracticalPlants page showing the last modification date of the plant.
     pub article_last_modified_at: Option<NaiveDateTime>,
+    ///
+    pub hardiness_zone: Option<String>,
```
This should be an int.

aidnurs replied:
These values are coming as partial ranges and partial numbers from the datasets, so we first implemented them as ranges, which worked pretty well on the PSQL side. However, there were some complications on the backend side. I was trying to figure out how the ranges should be imported in the backend for quite a while, but all my attempts were unsuccessful. The main problem is that diesel doesn't know how to interpret the ranges coming from the PSQL without telling him how to parse such values in `FromSQL` and `ToSQL` constructions. So I thought that sanitized and clean entries are currently more valuable for the project and could be adapted by backend experts in the future

markus2330 commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,230 @@ use diesel::{Identifiable, Insertable, Queryable};
+    /// The list of enum values.
+    pub propagation_method: Option<Vec<Option<PropagationMethod>>>,
+    ///
+    pub alternate_name: Option<String>,
```
Is this an alternate Latin name? Please add more comments.

aidnurs replied:
I'm actually not sure what is the difference between the `name`, which is used during the merge with the `common_name_en`, and the `alternate_name` column. Both value sets are coming from permapeople but have different values. Maybe Yvonne can provide us with some insights into these values once she sends me the biological part of the descriptions for the columns

markus2330 commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,230 @@ use diesel::{Identifiable, Insertable, Queryable};
+    ///
+    pub introduced_into: Option<String>,
+    ///
+    pub layer: Option<String>,
```
This conflicts with our terminology.

markus2330 reviewed (changes requested):
Looks like this was a lot of work, well done! :heart: I couldn't review everything today.

We are missing all the CSV files so that we can check correctness. Can you please put them all into the Database folder, moving all the old ones in a sub-folder with their date. Please also provide a CSV for Yvonne to check your comments. And also provide a fully merged CSV containing everything. Please add a README.md there so that we know which file is what, how it is generated, how @lukashartl can import it etc.

markus2330 commented:
CI also isn't happy at the moment.

markus2330 reviewed (changes requested):
We looked now over column names, which was a bit hard without up-to-date CSV but we did our best that you can finish soon. Major problem is that types are not as required: Columns we use within the app must be enum/int etc. and **not** String. We wrote it down where we saw problems.

I don't think we can do another review today as guests come over for my birthday. :balloon:

aidnurs commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,230 @@ use diesel::{Identifiable, Insertable, Queryable};
+    /// Family of the plant.
     pub family: Option<String>,
+    /// Subfamily of the plant.
     pub subfamily: Option<String>,
```
Ok, I thought you would also want to populate these values manually in the future. Removed

aidnurs commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,230 @@ use diesel::{Identifiable, Insertable, Queryable};
+    /// Subfamily of the plant.
     pub subfamily: Option<String>,
+    /// Genus of the plant.
     pub genus: Option<String>,
```
Yes, we don't have these relations on the code level yet, so it would be better to update the description when it is fixed and merged to the master in the next PRs

aidnurs commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,230 @@ use diesel::{Identifiable, Insertable, Queryable};
+    /// The list of enum values.
+    pub growth_rate: Option<Vec<Option<GrowthRate>>>,
+    ///
+    pub height: Option<String>,
+    ///
+    pub width: Option<String>,
```
The problem with width and height is that they are not fixed as number values across all datasets, there are many custom values or even ranges like `ft`, `feet`, `<number>m`, `<number> m`, etc. which is almost a manual rule set. I can only suggest leaving them as they are, but we can still define them all separately of course

aidnurs commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,230 @@ use diesel::{Identifiable, Insertable, Queryable};
+    ///
+    pub resistance: Option<String>,
+    ///
+    pub root_type: Option<String>,
```
There are more values. It is my bad that I was assuming you will be looking through the database itself instead of the `plants.csv`, because in comparison to the external dataset files this one was used only for presentation purposes. I'll update it incrementally from now on as well!

aidnurs commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,230 @@ use diesel::{Identifiable, Insertable, Queryable};
+    ///
+    pub habitat: Option<String>,
+    ///
+    pub spacing_en: Option<String>,
```
My last finding during fixing the backend was that the order of the columns should conform to the order of the schema.rs generated by diesel. I'll check which possible solutions we have here

markus2330 replied:
`spacing` (without _en) and Vec from int is missing. But it is only few values, so let us forget them.

aidnurs commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,230 @@ use diesel::{Identifiable, Insertable, Queryable};
+    ///
+    pub spacing_en: Option<String>,
+    ///
+    pub wikipedia: Option<String>,
```
The same problem with the un-updated plants.csv file. Column is renamed

aidnurs commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,230 @@ use diesel::{Identifiable, Insertable, Queryable};
+    /// The list of enum values.
+    pub growth_rate: Option<Vec<Option<GrowthRate>>>,
+    ///
+    pub height: Option<String>,
```
the same problem described for width column

aidnurs commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,230 @@ use diesel::{Identifiable, Insertable, Queryable};
+    ///
+    pub seed_viability: Option<String>,
+    ///
+    pub slug: Option<String>,
```
This value is the last part of the URL on permapeople webpage, however, in order to construct the external URL we need to access all the parent entities given by the `parent_id` column e.g. `https://permapeople.org/plants/citrullus-lanatus-watermelon/blacktail-1`, which is possible of course but not sure how deep this hierarchy can go on their webpage

aidnurs commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,230 @@ use diesel::{Identifiable, Insertable, Queryable};
+    ///
+    pub layer: Option<String>,
+    ///
+    pub leaves: Option<String>,
```
Excellent finding!

aidnurs commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,230 @@ use diesel::{Identifiable, Insertable, Queryable};
     pub genus: Option<String>,
-    pub edible_uses: Option<String>,
+    ///
+    pub edible_uses_en: Option<String>,
```
This is a bit tricky since there is no fixed structure and hierarchy of categories(there 3 levels of them) on the Reinsaat. For example, `New varieties 2023` are on the same level as `Beans`, and the same case is there for `Popcorn`, `Annual`, and `Cucumbers`. We could also look for Popcorn and Oil specifically within the values of course but this will add complexity to the code

aidnurs commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,230 @@ use diesel::{Identifiable, Insertable, Queryable};
+    /// The external url provided by the origin source.
+    pub external_url: Option<String>,
+    ///
+    pub root_depth: Option<String>,
```
`root_zone_tendency` is already merged. The same problem as for many other columns is valid here for `root_depth`, different value formats incl. strings, pure numbers, mixed, mixed measurement units, etc. So I was following the same thought process here - seeking for the database completeness

aidnurs commented:
Hey @markus2330,

thank you for your feedback! To conclude all the recent updates:
- all CSV files are updated on the Nextcloud
- the main challenge in defining enums and data types is the diversity of the values across all datasets. Since the values were not fixed by any data type on the external sources, their users were free to submit anything they want. I've applied as many data types as possible, I've also defined custom rule sets for the fields to sanitize and unify value formats where it made sense. However, there were still some fields where this transformation was not reasonable anymore, because these would require a set translating every row differently
- updated technical documentation for backend structs
- updated code documentation
- prepared squashed migrations for the later usage

Please let me know if there is anything else missing.

Sincerely,
Nursultan

markus2330 commented on backend/schema_dump/2023-04-23-193141_create_plants/up.sql:
```diff
@@ -0,0 +1,176 @@
+    updated_at TIMESTAMP DEFAULT now() NOT NULL,
+    has_drought_tolerance BOOLEAN,
+    tolerates_wind BOOLEAN,
+    plant_references TEXT [],
```
Better to not repeat the table name:

```suggestion
    references TEXT [],
```

markus2330 commented on backend/schema_dump/2023-04-23-193141_create_plants/up.sql:
```diff
@@ -0,0 +1,176 @@
+    alternate_name TEXT,
+    diseases TEXT,
+    edible BOOLEAN,
+    edible_parts TEXT [],
```
Is okay as text, enum would be preferable (to have suggestions in advanced search)

markus2330 commented on backend/schema_dump/2023-04-23-193141_create_plants/up.sql:
```diff
@@ -0,0 +1,176 @@
+    diseases TEXT,
+    edible BOOLEAN,
+    edible_parts TEXT [],
+    germination_temperature TEXT,
```
We agree it would be too much work for you to bring this to a range of int, so let us leave it.

markus2330 commented on backend/schema_dump/2023-04-23-193141_create_plants/up.sql:
```diff
@@ -0,0 +1,176 @@
+    propagation_transplanting TEXT,
+    resistance TEXT,
+    root_type TEXT,
+    seed_planting_depth_en TEXT,
```
```suggestion
    seed_planting_depth TEXT,
```

aidnurs replied:
The naming was suggested in the hierarchy file, there is also the corresponding _de column. Let me know if this is still the case or if I should change it back to `seed_planting_depth`

markus2330 commented on backend/schema_dump/2023-04-23-193141_create_plants/up.sql:
```diff
@@ -0,0 +1,176 @@
+    seed_planting_depth_en TEXT,
+    seed_viability TEXT,
+    slug TEXT,
+    spread TEXT,
```
Should be ideally merged with width.

markus2330 commented on backend/schema_dump/2023-04-23-193141_create_plants/up.sql:
```diff
@@ -0,0 +1,176 @@
+    slug TEXT,
+    spread TEXT,
+    utility TEXT,
+    warning TEXT,
```
Enum would be marvelous.

markus2330 commented on backend/schema_dump/2023-04-23-193141_create_plants/up.sql:
```diff
@@ -0,0 +1,176 @@
+    spread TEXT,
+    utility TEXT,
+    warning TEXT,
+    when_to_plant_cuttings TEXT,
```
```suggestion
    when_to_plant_cuttings_en TEXT,
```

markus2330 commented on backend/schema_dump/2023-04-23-193141_create_plants/up.sql:
```diff
@@ -0,0 +1,176 @@
+    utility TEXT,
+    warning TEXT,
+    when_to_plant_cuttings TEXT,
+    when_to_plant_division TEXT,
+    when_to_plant_transplant TEXT,
```
```suggestion
    when_to_plant_division_en TEXT,
    when_to_plant_transplant_en TEXT,
```

markus2330 commented on backend/schema_dump/2023-04-23-193141_create_plants/up.sql:
```diff
@@ -0,0 +1,176 @@
+    when_to_plant_cuttings TEXT,
+    when_to_plant_division TEXT,
+    when_to_plant_transplant TEXT,
+    when_to_sow_indoors TEXT,
```
```suggestion
    when_to_sow_indoors_en TEXT,
```

markus2330 commented on backend/schema_dump/2023-04-23-193141_create_plants/up.sql:
```diff
@@ -0,0 +1,176 @@
+    when_to_sow_indoors TEXT,
+    sowing_outdoors_en TEXT,
+    when_to_start_indoors_weeks TEXT,
+    when_to_start_outdoors_weeks TEXT,
```
What is a negative number here?

aidnurs replied:
This is raw data coming from Permapeople. Since the same negative number occurs multiple times, it is probably done on purpose

markus2330 commented on backend/schema_dump/2023-04-23-193141_create_plants/up.sql:
```diff
@@ -0,0 +1,176 @@
+    when_to_start_outdoors_weeks TEXT,
+    cold_stratification_temperature TEXT,
+    cold_stratification_time TEXT,
+    days_to_harvest TEXT,
```
Should be vec of int. It might be possible to guess this number using sowing_outdoors and harvest_time but then Yvonne would need to check which rows have which of these columns. Should Yvonne check this?

markus2330 commented on backend/schema_dump/2023-04-23-193141_create_plants/up.sql:
```diff
@@ -0,0 +1,176 @@
+    habitat TEXT,
+    spacing_en TEXT,
+    wikipedia_url TEXT,
+    days_to_maturity TEXT,
```
could be int but days to harvest is more important.

markus2330 commented on backend/schema_dump/2023-04-23-193141_create_plants/up.sql:
```diff
@@ -0,0 +1,176 @@
+    wikipedia_url TEXT,
+    days_to_maturity TEXT,
+    pests TEXT,
+    version TEXT,
```
Should be positive smallint.

markus2330 commented on backend/schema_dump/2023-04-23-193141_create_plants/up.sql:
```diff
@@ -0,0 +1,176 @@
+    external_article_number TEXT,
+    external_portion_content TEXT,
+    sowing_outdoors_de TEXT,
+    sowing_outdoors SMALLINT [],
+    harvest_time SMALLINT [],
```
Great to have these two. But "Pre-cultivation"/"Voranzucht" is missing. greenhouse is optional.

markus2330 commented on backend/schema_dump/2023-04-23-193141_create_plants/up.sql:
```diff
@@ -0,0 +1,176 @@
+    seed_weight_1000_de TEXT,
+    seed_weight_1000_en TEXT,
+    seed_weight_1000 FLOAT,
+    machine_cultivation_possible TEXT,
```
This should be boolean.

markus2330 reviewed (changes requested):
Great work, we are getting to the details. We looked at all columns again. We tried our best to separate where better data types are a must and where it is nice to have. Please prioritize your work accordingly. As next step we will check for correctness of the scraper. Hopefully we won't find anything, then this PR could be finished within a day.

markus2330 reviewed (changes requested):
Great work, we are getting to the details. We looked at all columns again. We tried our best to separate where better data types are a must and where it is nice to have. Please prioritize your work accordingly. As next step we will check for correctness of the scraper. Hopefully we won't find anything, then this PR could be finished within a day.

markus2330 commented:
GitHub had some bug when submitting the large review but it looks like somehow it managed to show :laughing:

markus2330 commented on scraper/src/fetch_practicalplants.js:
```diff
@@ -1,76 +1,13 @@
-  return distinctFamilyDetails;
-}
-
-function createDistinctGenusDetails(details) {
```
Genus entries (column genus) are sometimes missing, they can be extracted from unique_name. Also here: entries for every genus should end up in plants.csv (and plants table).

aidnurs replied:
sure, I'll add this to the leftover ticket. I've noticed some individual problems caused by the hierarchy naming, still not sure if it is applicable to every plant in the database

markus2330 replied:
Hierarchy and enums/ints are the most important goals. I would like to avoid that someone else needs to dive into your scraper, which would take lots of time.

markus2330 commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,224 @@ use diesel::{Identifiable, Insertable, Queryable};
+    pub growth_rate: Option<Vec<Option<GrowthRate>>>,
+    /// Fetched from PracticalPlants as \`mature_size_height\` and merged with Permapeople. (fetched from PracticalPlants and Permapeople)
+    pub height: Option<String>,
+    /// Fetched from PracticalPlants as \`mature_size_width\` and merged with Permapeople. (fetched from PracticalPlants and Permapeople)
```
Units must be mentioned here (cm).
```suggestion
    /// Unit: Centimeter, Fetched from PracticalPlants as \`mature_size_width\` and merged with Permapeople. (fetched from PracticalPlants and Permapeople)
```

aidnurs replied:
I'm not sure if all the values have the same units e.g. there are `feet` values inside. the same challenge I described for the diversity of value formatting.

markus2330 commented on backend/src/model/entity.rs:
```diff
@@ -11,70 +11,224 @@ use diesel::{Identifiable, Insertable, Queryable};
+    /// (fetched from Reinsaat)
+    pub seed_planting_depth_de: Option<String>,
+    /// (fetched from Reinsaat)
+    pub seed_weight_1000_de: Option<String>,
+    /// (fetched from Reinsaat)
+    pub seed_weight_1000_en: Option<String>,
+    /// Fetched from Permapeople as \`1000_seed_weight_g\` and renamed. (fetched from Permapeople)
+    pub seed_weight_1000: Option<f64>,
```
For some plants some or all of them are missing, e.g. https://www.reinsaat.at/shop/DE/paprika/spitzpaprika_1/kurtovska_kapiya_1/
Even though they are present (Tausendkorngewicht (TKG): 7.15 g)

Other example: https://www.reinsaat.at/shop/DE/paprika/spitzpaprika_1/shintokiwa/ does not have `seed_weight_100` even though present Tausendkorngewicht (TKG): 24.54 g (but has _de and _en)

markus2330 reviewed (changes requested):
Hopefully last review! Really amazing work. Keep your head up for the last fixes :rocket:

aidnurs commented on frontend/package-lock.json:
```diff
@@ -6,7 +6,7 @@
     "": {
       "name": "permaplant",
-      "version": "0.0.0",
+      "version": "0.1.0",
```
during the `npm install` step, a version mentioned in the package.json is replicated to the package-lock file, so it stays consistent. so this change wasn't done manually by me, it just took over the package version

aidnurs commented on scraper/src/helpers/column_mapping_permapeople.js:
```diff
@@ -0,0 +1,229 @@
+ * column names of the database.
+ *
+ * Key is the permapeople column name.
+ * Value is an object with the following properties:
```
Good point, updated the comment!

aidnurs replied:
Nice suggestion in general, I will go through the comments again and try to describe all the possible cases

aidnurs commented on scraper/doc/practicalplants_scraper.md:
```diff
@@ -0,0 +1,62 @@
+
+Please note the following points when correcting the data:
+
+- The `Binomial name` column should not be changed.
```
I kept these namings on purpose because I store all the column names as they are during the fetch step from origin sources, and modify/merge them later during the merge step. However, the good catch here, I should update the documentation stating this

aidnurs commented on backend/schema_dump/2023-04-23-193141_create_plants/up.sql:
```diff
@@ -0,0 +1,176 @@
+    life_cycle LIFE_CYCLE [],
+    growth_rate GROWTH_RATE [],
+    height TEXT,
+    width TEXT,
```
For the case of hardiness zones vector of int values is a great solution, however, here we have the same problem as for other custom input columns i.e. diversity of the values e.g. `..cm`,`..m`, `..inches`, range of numbers, etc. Please let me know what you think about such inputs

aidnurs commented on scraper/src/insert_plants.js:
```diff
@@ -0,0 +1,121 @@
+  const cs = new pgp.helpers.ColumnSet(dbPlantsColumns, {
+    table: 'plants',
+  });
+
```
I use the upserts here so the data just gets updated if something is changed. I cannot imagine a case when a plant will not exist anymore, so I would not suggest cleaning the table before the insertion. This way we always have at least as many entries as all the data sources.

markus2330 reviewed (changes requested):
Unfortunately, I get an error, also after resetting the database. Please also add  mergedDatasets.csv (that works) to Nextcloud.


```
[INFO] Plant details inserted into database.
/home/markus/Projekte/PermaplanT/scraper/node_modules/pg-protocol/dist/parser.js:287
        const message = name === 'notice' ? new messages_1.NoticeMessage(length, messageValue) : new messages_1.DatabaseError(messageValue, length, name);
                                                                                                 ^

error: Befehl in ON CONFLICT DO UPDATE kann eine Zeile nicht ein zweites Mal ändern
    at Parser.parseErrorMessage (/home/markus/Projekte/PermaplanT/scraper/node_modules/pg-protocol/dist/parser.js:287:98)
    at Parser.handlePacket (/home/markus/Projekte/PermaplanT/scraper/node_modules/pg-protocol/dist/parser.js:126:29)
    at Parser.parse (/home/markus/Projekte/PermaplanT/scraper/node_modules/pg-protocol/dist/parser.js:39:38)
    at Socket.<anonymous> (/home/markus/Projekte/PermaplanT/scraper/node_modules/pg-protocol/dist/index.js:11:42)
    at Socket.emit (node:events:513:28)
    at addChunk (node:internal/streams/readable:324:12)
    at readableAddChunk (node:internal/streams/readable:297:9)
    at Readable.push (node:internal/streams/readable:234:10)
    at TCP.onStreamRead (node:internal/stream_base_commons:190:23) {
  length: 299,
  severity: 'FEHLER',
  code: '21000',
  detail: undefined,
  hint: 'Stellen Sie sicher, dass keine im selben Befehl fürs Einfügen vorgesehene Zeilen doppelte Werte haben, die einen Constraint verletzen würden.',
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'nodeModifyTable.c',
  line: '1665',
  routine: 'ExecOnConflictUpdate',
  query: `insert into "plants"("common_name_en","common_name_de","unique_name","genus","family","edible_uses_en","medicinal_uses","material_uses_and_functions","botanic","material_uses","functions","heat_zone","shade","soil_ph","soil_texture","soil_water_retention","environmental_tolerances","native_geographical_range","native_environment","ecosystem_niche","deciduous_or_evergreen","herbaceous_or_woody","life_cycle","growth_rate","height","width","fertility","flower_colour","flower_type","has_drought_tolerance","tolerates_wind","plant_references","is_tree","nutrition_demand","article_last_modified_at","hardiness_zone","light_requirement","water_requirement","propagation_method","alternate_name","diseases","edible","edible_parts","germination_temperature","introduced_into","habitus","medicinal_parts","native_to","plants_for_a_future","plants_of_the_world_online_link","plants_of_the_world_online_link_synonym","pollination","propagation_transplanting_en","resistance","root_type","seed_planting_depth_en","seed_viability","slug","spread","utility","warning","when_to_plant_cuttings_en","when_to_plant_division_en","when_to_plant_transplant_en","when_to_sow_indoors_en","sowing_outdoors_en","when_to_start_indoors_weeks","when_to_start_outdoors_weeks","cold_stratification_temperature","cold_stratification_time","days_to_harvest","habitat","spacing_en","wikipedia_url","days_to_maturity","pests","version","germination_time","description","parent_id","external_source","external_id","external_url","root_depth","external_article_number","external_portion_content","sowing_outdoors_de","sowing_outdoors","harvest_time","spacing_de","required_quantity_of_seeds_de","required_quantity_of_seeds_en","seed_planting_depth_de","seed_weight_1000_de","seed_weight_1000_en","seed_weight_1000","machine_cultivation_possible","edible_uses_de") values(array['indian abelia']::text[],null::text[],'Abelia triflora','Abelia','Caprifoliaceae',null,null,'[1][2]','[3][6]',null,null,null,'light shade',null::soil_ph[],array['sandy','loamy']::soil_texture[],null::soil_water_retention[],null::text[],null,null,null,'deciduous'::deciduous_or_evergreen,'woody'::herbaceous_or_woody,array['perennial']::life_cycle[],array['moderate']::growth_rate[],'3.5','3',null::fertility[],null,'hermaphrodite'::flower_type,'false'::boolean,'false',array['Gamble. J. S. A Manual of Indian Timbers. Bishen Singh Mahendra Pal Singh  (32202/01/01)','Gupta. B. L. Forest Flora of Chakrata, Dehra Dun and Saharanpur. Forest Research Institute Press  (32202/01/01)','Bean. W. Trees and Shrubs Hardy in Great Britain. Vol 1 - 4 and Supplement. Murray  (32202/01/01)','Sheat. W. G. Propagation of Trees, Shrubs and Conifers. MacMillan and Co  (32202/01/01)','Genders. R. Scented Flora of the World. Robert Hale. London. ISBN 0-7090-5440-8 (32202/01/01)','Huxley. A. The New RHS Dictionary of Gardening. 1992. MacMillan Press ISBN 0-333-47494-5 (32202/01/01)','Grey-Wilson. C. &amp; Matthews. V. Gardening on Walls Collins ISBN 0-00-219220-0 (32202/01/01)','Davis. B. Climbers and Wall Shrubs. Viking. ISBN 0-670-82929-3 (32202/01/01)','Phillips. R. &amp; Rix. M. Shrubs. Pan Books ISBN 0-330-30258-2 (32202/01/01)','Thomas. G. S. Ornamental Shrubs, Climbers and Bamboos. Murray ISBN 0-7195-5043-2 (32202/01/01)']::text[],null,null::nutrition_demand,'2013-05-04T09:57:00.000Z','[5,9)',array['full sun','partial sun/shade']::light_requirement[],array['dry','moist']::water_requirement[],null::propagation_method[],null,null,null::boolean,null::text[],null,null,'Shrubs',null,'Afghanistan, China South-Central, Nepal, Pakistan, Tibet, West Himalaya','https://pfaf.org/User/Plant.aspx?LatinName=Abelia triflora','https://powo.science.kew.org/taxon/urn:lsid:ipni.org:names:148232-1','https://powo.science.kew.org/taxon/urn:lsid:ipni.org:names:150126-1',null,null,null,null,null,null,'abelia-triflora-indian-abelia',null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,'8',null,null,null,'permapeople','8406',null,null,null,null,null,null::integer[],null::integer[],null,null,null,null,null,null,null,null::boolean,null),(array['okra']::text[],null::text[],'Abelmoschus esculentus','Abelmoschus','Malvaceae','Oil',null,'[11][12][4][13][14][15][16]','[10]',null,null,null,'no shade',null::soil_ph[],array['sandy','loamy','clay']::soil_texture[],array['well drained']::soil_water_retention[],null::text[],null,null,null,null::deciduous_or_evergreen,null::herbaceous_or_woody,array['annual','perennial']::life_cycle[],null::growth_rate[],'1.0',null,null::fertility[],null,'hermaphrodite'::flower_type,'false'::boolean,'false',array['Hedrick. U. P. Sturtevant''s Edible Plants of the World. Dover Publications ISBN 0-486-20459-6 (32202/01/01)','Vilmorin. A. The Vegetable Garden. Ten Speed Press ISBN 0-89815-041-8 ()','Rice. G. (Editor) Growing from Seed. Volume 1. Thompson and Morgan.  (32202/01/01)','Komarov. V. L. Flora of the USSR. Israel Program for Scientific Translation  (32202/01/01)','Facciola. S. Cornucopia - A Source Book of Edible Plants. Kampong Publications ISBN 0-9628087-0-9 (32202/01/01)','Harris. B. C. Eat the Weeds. Pivot Health  (32202/01/01)','Kunkel. G. Plants for Human Consumption. Koeltz Scientific Books ISBN 3874292169 (32202/01/01)','Chopra. R. N., Nayar. S. L. and Chopra. I. C. Glossary of Indian Medicinal Plants (Including the Supplement). Council of Scientific and Industrial Research, New Delhi.  (32202/01/01)','Cribb. A. B. and J. W. Wild Food in Australia. Fontana ISBN 0-00-634436-4 (32202/01/01)','Huxley. A. The New RHS Dictionary of Gardening. 1992. MacMillan Press ISBN 0-333-47494-5 (32202/01/01)','Schery. R. W. Plants for Man.   ()','Usher. G. A Dictionary of Plants Used by Man. Constable ISBN 0094579202 (32202/01/01)','Buchanan. R. A Weavers Garden.   ()','Hill. A. F. Economic Botany. The Maple Press  (32202/01/01)','Bell. L. A. Plant Fibres for Papermaking. Liliaceae Press  (32202/01/01)','Stuart. Rev. G. A. Chinese Materia Medica. Taipei. Southern Materials Centre  ()','Grieve. A Modern Herbal. Penguin ISBN 0-14-046-440-9 (32202/01/01)','Lust. J. The Herb Book. Bantam books ISBN 0-553-23827-2 (32202/01/01)','Manandhar. N. P. Plants and People of Nepal Timber Press. Oregon. ISBN 0-88192-527-6 (32202/01/01)','Phillips. R. &amp; Rix. M. Vegetables Macmillan Reference Books, London. ISBN 0 333 62640 0 (32202/01/01)','Phillips. R. &amp; Rix. M. Conservatory and Indoor Plants Volumes 1 &amp; 2 Pan Books, London. ISBN 0-330-37376-5 (32202/01/01)']::text[],null,null::nutrition_demand,'2012-12-26T14:33:00.000Z','[5,11)',array['full sun']::light_requirement[],array['moist']::water_requirement[],null::propagation_method[],'A-koto, angu, apala, asowntem, bakhua-mun, bamia, bandakka, bendi, bhindee, bhindi, binda, bindi, bondo, cantarela, derere rechipudzi, derere, dheras, dherosh, enmomi, fetri, gombaut, gombo, gumbo, guro, gusha, hakuyot, idelele, ikhievbo, ilasha, ilo, ka fei huang kui, kacang bendi, kaganh lender, kandia, kandjie, kopi arab, krachiap-mon, kubewa, lafeu, lieka, loka, maana, ma-lontho, mesta, muomi, miagorro, nathando, nkruma, obori, ochro, okworu, okwulu, otigo-iwoka, pahari bendi, pingpesi, poot barang, pui, quiabo, quimbambo, saluyot a bunga, sayur bendi, taku, uisul hme, vandakai, vandikkai, vendal, wayika, you-padi',null,'true'::boolean,array['fruit','leaves','root','seed']::text[],null,'Alabama, Albania, Andaman Is., Angola, Assam, Bahamas, Benin, Borneo, Bulgaria, Burkina, Cambodia, Cape Verde, Caroline Is., Cayman Is., Central African Repu, Chad, China South-Central, China Southeast, Comoros, Congo, Cuba, Dominican Republic, East Himalaya, Eritrea, Fiji, Florida, Gabon, Gambia, Georgia, Greece, Guinea-Bissau, Gulf of Guinea Is., Hainan, Haiti, Illinois, Ivory Coast, Jamaica, Jawa, Krym, KwaZulu-Natal, Laos, Leeward Is., Louisiana, Malaya, Mali, Marianas, Marshall Is., Maryland, Mauritania, Mexico Southwest, Mississippi, Mozambique, Nicobar Is., Niger, Nigeria, North Carolina, Northern Provinces, Oman, Pakistan, Peru, Philippines, Puerto Rico, Romania, Senegal, Sierra Leone, South Carolina, South European Russi, Southwest Caribbean, Sri Lanka, Sudan, Tanzania, Thailand, Togo, Uganda, Ukraine, Venezuela, Venezuelan Antilles, Vietnam, Virginia, Windward Is., Zambia, Zaïre, Zimbabwe','Herbs',null,'Bangladesh, India, Myanmar','https://pfaf.org/User/Plant.aspx?LatinName=Abelmoschus esculentus','https://powo.science.kew.org/taxon/urn:lsid:ipni.org:names:558006-1',null,null,null,null,null,null,null,'abelmoschus-esculentus-okra',null,null,'Weed potential',null,null,null,null,null,null,null,null,null,null,null,null,'https://en.wikipedia.org/wiki/Okra',null,null,'16',null,null,null,'permapeople','233',null,null,null,null,null,null::integer[],null::integer[],null,null,null,null,null,null,null,null::boolean,null),(array['aibika']::text[],array['maniok-bisameibisch']::text[],'Abelmoschus manihot','Abelmoschus','Malvaceae',null,null,null,'[2][6]',null,null,null,'no shade',null::soil_ph[],array['sandy','loamy','clay']::soil_texture[],array['well drained']::soil_water_retention[],null::text[],null,null,null,null::deciduous_or_evergreen,null::herbaceous_or_woody,array['perennial']::life_cycle[],array['vigorous']::growth_rate[],'3.0',null,null::fertility[],null,'hermaphrodite'::flower_type,'false'::boolean,null,array['Facciola. S. Cornucopia - A Source Book of Edible Plants. Kampong Publications ISBN 0-9628087-0-9 (1990-00-00)','Huxley. A. The New RHS Dictionary of Gardening. 1992. MacMillan Press ISBN 0-333-47494-5 (1992-00-00)','Chopra. R. N., Nayar. S. L. and Chopra. I. C. Glossary of Indian Medicinal Plants (Including the Supplement). Council of Scientific and Industrial Research, New Delhi.  (1986-00-00)','Manandhar. N. P. Plants and People of Nepal Timber Press. Oregon. ISBN 0-88192-527-6 (2002-00-00)','Phillips. R. &amp; Rix. M. Conservatory and Indoor Plants Volumes 1 &amp; 2 Pan Books, London. ISBN 0-330-37376-5 (1998-00-00)','[Flora of China]  (1994-00-00)']::text[],null,`... 20677554 more characters,
  params: undefined
}

Node.js v19.9.0
```
